# Eden Anderson
## Module 4 Assignment / 44620

# GitHub Repository Link



# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Eden Anderson

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

# Question 1

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [1]:
import requests
import json

result = json.loads(requests.get('https://lyrist.vercel.app/api/exile/taylorswift').text)
print(result)

with open('result.json', 'w') as f:
    json.dump(result, f)

{'title': 'Exile', 'artist': 'Taylor Swift', 'lyrics': "I can see you standing, honey\nWith his arms around your body\nLaughin', but the joke's not funny at all\n\nAnd it took you five whole minutes\nTo pack us up and leave me with it\nHoldin' all this love out here in the hall\n\nI think I've seen this film before\nAnd I didn't like the ending\nYou're not my homeland anymore\nSo what am I defending now?\nYou were my town\nNow I'm in exile, seein' you out\nI think I've seen this film before\n\nI can see you starin', honey\nLike he's just your understudy\nLike you'd get your knuckles bloody for me\n\nSecond, third, and hundredth chances\nBalancin' on breaking branches\nThose eyes add insult to injury\n\nI think I've seen this film before\nAnd I didn't like the ending\nI'm not your problem anymore\nSo who am I offending now?\nYou were my crown\nNow I'm in exile, seein' you out\nI think I've seen this film before\nSo I'm leavin' out the side door\n\nSo step right out, there is no amount\n

# Question 2

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [2]:
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

f = open('result.json')
  
data = json.load(f)

print(data.get('lyrics'))

nlp = spacy.load("en_core_web_sm")
nlp.add_pipe('spacytextblob')
text = data.get('lyrics')
doc = nlp(text)

print("The polarity score of the sentiment analysis is ", doc._.blob.polarity)

I can see you standing, honey
With his arms around your body
Laughin', but the joke's not funny at all

And it took you five whole minutes
To pack us up and leave me with it
Holdin' all this love out here in the hall

I think I've seen this film before
And I didn't like the ending
You're not my homeland anymore
So what am I defending now?
You were my town
Now I'm in exile, seein' you out
I think I've seen this film before

I can see you starin', honey
Like he's just your understudy
Like you'd get your knuckles bloody for me

Second, third, and hundredth chances
Balancin' on breaking branches
Those eyes add insult to injury

I think I've seen this film before
And I didn't like the ending
I'm not your problem anymore
So who am I offending now?
You were my crown
Now I'm in exile, seein' you out
I think I've seen this film before
So I'm leavin' out the side door

So step right out, there is no amount
Of crying I can do for you
All this time
We always walked a very thin line
You didn't even

Question 2 - Answered Question

With a polarity score that is positive, it shows that sentiment analysis believes the lyrics have more of a positive connotation; with the score not being too much above zero, however, the lyrics still are more neutral than a more positive score.  I am honestly a bit surprised that the score landed on the positive side, since the song has so many "never"-type words used. (Plus the melody is more melancholoy, but I have that as an added advantage for context.)

# Question 3

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [10]:
# getting input for function
artist = input("Name of Artist (no spaces): ")
song = input("Name of Song (no spaces): ")
filename = input("Name of the file (add.json): ")

def songs():
    lyrics_dict = requests.get('https://lyrist.vercel.app/api/{:}/{:}'.format(artist, song)).json()
    lyrics = lyrics_dict.get('lyrics')
    print("\n", "Lyrics...", "\n", lyrics)
    
with open(filename, 'w') as f:
    json.dump(lyrics, f)

songs()

Name of Artist (no spaces):  taylorswift
Name of Song (no spaces):  fearless
Name of the file (add.json):  Fearless.json



 Lyrics... 
 There's somethin' bout the way
The street looks when it's just rained
There's a glow off the pavement
You walk me to the car
And you know I wanna ask you to dance right there
In the middle of the parking lot
Yeah
Oh, yeah

We're drivin' down the road
I wonder if you know
I'm tryin' so hard not to get caught up now
But you're just so cool
Run your hands through your hair
Absent mindedly makin' me want you

And I don't know how it gets better than this
You take my hand and drag me head first
Fearless
And I don't know why
But with you I'd dance in a storm
In my best dress
Fearless

So baby drive slow
'Til we run out of road in this one horse town
I wanna stay right here in this passenger seat
You put your eyes on me
In this moment now capture it, remember it

'Cause I don't know how it gets better than this
You take my hand and drag me head first
Fearless
And I don't know why
But with you I'd dance in a storm
In my best dress
Fearless
Oh, oh

Well you stood there with me in 

# Question 4

4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [13]:
## getting input for function

filename = input("Name of the file: ")

f = open(filename)

nlp = spacy.load("en_core_web_sm")
nlp.add_pipe('spacytextblob')
text = data = json.load(f)
doc = nlp(text)

print("The polarity score of the sentiment analysis for this song is ", doc._.blob.polarity)

Name of the file:  Fearless.json


The polarity score of the sentiment analysis for this song is  0.3153902116402117


Question 4 - Answered Question

Here are 4 of the polarity answers from saved files.  (Song titles in file names)

1:  Name of the file:  BlankSpace.json
The polarity score of the sentiment analysis for this song is  -0.046432900432900416

2:  Name of the file:  22.json
The polarity score of the sentiment analysis for this song is  0.15178571428571427

3:  Name of the file:  Enchanted.json
The polarity score of the sentiment analysis for this song is  0.2446078431372549

4:  Name of the file:  Fearless.json
The polarity score of the sentiment analysis for this song is  0.3153902116402117

I believe the polarity of the song lyrics matches the tone of they lyrics better than I expected, after saving the first file.  This is especially noticeable in examples 3 and 4, which are more positive songs - with positive messages.  I also was excited to see the negative polarity, just to have an example of that, but also because it better recognizes some of the bleakness of the lyrics in "Blank Space."  I still believe that melody plays a bigger part than just the lyrics of the song, as exemplified in the song "22," which makes the listener feel more positivity than just the lyrics alone might, if left to sentiment analysis alone.

